clippy_lints/src/use_self.rs

```diff
@@ -226,7 +226,12 @@ impl<'tcx> LateLintPass<'tcx> for UseSelf {
             } else {
                 hir_ty_to_ty(cx.tcx, hir_ty)
             }
-            && same_type_and_consts(ty, cx.tcx.type_of(impl_id).instantiate_identity())
+            && let impl_ty = cx.tcx.type_of(impl_id).instantiate_identity()
+            && same_type_and_consts(ty, impl_ty)
+            // Ensure the type we encounter and the one from the impl have the same lifetime parameters. It may be that
+            // the lifetime parameters of `ty` are ellided (`impl<'a> Foo<'a> { fn new() -> Self { Foo{..} } }`, in
+            // which case we must still trigger the lint.
+            && (same_lifetimes(ty, impl_ty) || has_no_lifetime(ty))
```

Comment:
`has_no_lifetime` makes less operations so I think it'd be better to have it run first.

```suggestion
            && (has_no_lifetime(ty) || same_lifetimes(ty, impl_ty))
```

Reply:
Me neither so let's wait for a clippy reviewer to see waht they prefer. :)

Reply:
Sorry for the delay. In general, it's better to optimistically check special cases before the complex general case.
So checking for no lifetimes would be better.